### ISSUE TITLE
OaepEncoding with Sha256 Digest is not decipherable with OpenSSL

### ISSUE THREAD

**Issue by sturcotte06** [bcgit/bc-csharp#161]
```csharp
using System.Text;
using System.IO;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Digests;
using Org.BouncyCastle.Crypto.Encodings;
using Org.BouncyCastle.Crypto.Engines;
using Org.BouncyCastle.Crypto.Generators;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Security;
using Org.BouncyCastle.OpenSsl;

namespace Example
{
    class Program
    {
        public static void Main()
        {
            var generator = new RsaKeyPairGenerator();
            generator.Init(new KeyGenerationParameters(new SecureRandom(), 2048));

            var keyPair = generator.GenerateKeyPair();
            
            using (var f = File.Open("key.ppk", FileMode.OpenOrCreate))
            {
                using (TextWriter textWriter = new StreamWriter(f))
                {
                    PemWriter wr = new PemWriter(textWriter);
                    wr.WriteObject(keyPair.Private);
                }
            }

            using (var f = File.Open("key.pub", FileMode.OpenOrCreate))
            {
                using (TextWriter textWriter = new StreamWriter(f))
                {
                    PemWriter wr = new PemWriter(textWriter);
                    wr.WriteObject(keyPair.Public);
                }
            }

            var payload = "Hello World!";
            var payloadBytes = Encoding.ASCII.GetBytes(payload);
            var keyParams = (RsaKeyParameters) keyPair.Public;

            var cipher = new OaepEncoding(new RsaEngine(), new Sha256Digest());
            cipher.Init(true, keyParams);

            var encrypted = cipher.ProcessBlock(payloadBytes, 0, payloadBytes.Length);

            using (var f = File.Open("encrypted.bin", FileMode.OpenOrCreate))
            {
                f.Write(encrypted, 0, encrypted.Length);
            }
        }
    }
}
```

When trying to decrypt the string with openssl:
```bash
user@host:~$ openssl rsautl -decrypt -inkey key.ppk -in encrypted.bin -oaep
RSA operation error
140659123360192:error:04099079:rsa routines:RSA_padding_check_PKCS1_OAEP_mgf1:oaep decoding error:../crypto/rsa/rsa_oaep.c:240:
140659123360192:error:04065072:rsa routines:rsa_ossl_private_decrypt:padding check failed:../crypto/rsa/rsa_ossl.c:487:
```

Replacing 
```csharp
var cipher = new OaepEncoding(new RsaEngine(), new Sha256Digest());
``` 
with 
```csharp
var cipher = new OaepEncoding(new RsaEngine(), new Sha1Digest());
```
works as intended.

**Comment by tommed** [comment]
This is a pretty big bug for us as we want to use GCP's KMS service to perform verification services between various micro-services and .NET (using BC) is the only platform we're unable to match the RSA/SHA256/Oaep profile that KMS uses. We can't use SHA1 as it's not on the supported list (as it's not strong enough).

**Comment by izydorkowalsky** [comment]
OpenSSL's rsautil seems to implement just SHA1 for OAEP-Padding.
Did you try with OpenSSL's cms or pkeyutil ???
https://superuser.com/questions/1016696/using-a-hash-other-than-sha1-for-oaep-with-openssl-cli

**Comment by dghgit** [comment]
Closing this one as if there's a bug anywhere it's not ours. Either that or we've spent the last 5 years validating with NIST and both us and them have been missing something. One thing - in the example posted -oaep by default is SHA-1 with MGF1 using SHA-1, this is what the working code uses. To specify SHA-256 there would have to be an additional parameter to the openssl command as there's nothing in the encrypted blob to tell it to use SHA-256. That information is only available in the algorithm parameters.
